File: src/main/VHD Director/ExceptionHandling_demo/ClassLibrary1/HandledExceptionManager.vb
Imports System.Windows.Forms
Imports System.Threading
Imports System.Text.RegularExpressions
Imports System.Configuration

'--
'-- Generic HANDLED error handling class
'--
'-- It's like MessageBox, but specific to handled exceptions, and supports email notifications
'--
'-- Jeff Atwood
'-- http://www.codinghorror.com
'--
Public Class HandledExceptionManager

    Private Shared _blnHaveException As Boolean = False
    Private Shared _blnEmailError As Boolean = True
    Private Shared _strEmailBody As String
    Private Shared _strExceptionType As String
    Private Const _strDefaultMore As String = "No further information is available. If the problem persists, contact (contact)."

    Public Shared Property EmailError() As Boolean
        Get
            Return _blnEmailError
        End Get
        Set(ByVal Value As Boolean)
            _blnEmailError = Value
        End Set
    End Property

    Public Enum UserErrorDefaultButton
        [Default] = 0
        Button1 = 1
        Button2 = 2
        Button3 = 3
    End Enum

    '-- 
    '-- replace generic constants in strings with specific values
    '--
    Private Shared Function ReplaceStringVals(ByVal strOutput As String) As String
        Dim strTemp As String
        If strOutput Is Nothing Then
            strTemp = ""
        Else
            strTemp = strOutput
        End If
        strTemp = strTemp.Replace("(app)", AppSettings.AppProduct)
        strTemp = strTemp.Replace("(contact)", AppSettings.GetString("UnhandledExceptionManager/ContactInfo"))
        Return strTemp
    End Function

    '--
    '-- make sure "More" text is populated with something useful
    '--
    Private Shared Function GetDefaultMore(ByVal strMoreDetails As String) As String
        If strMoreDetails = "" Then
            Dim objStringBuilder As New System.Text.StringBuilder
            With objStringBuilder
                .Append(_strDefaultMore)
                .Append(Environment.NewLine)
                .Append(Environment.NewLine)
                .Append("Basic technical information follows: " + Environment.NewLine)
                .Append("---" + Environment.NewLine)
                .Append(UnhandledExceptionManager.SysInfoToString(True))
            End With
            Return objStringBuilder.ToString
        Else
            Return strMoreDetails
        End If
    End Function

    '--
    '-- converts exception to a formatted "more" string
    '--
    Private Shared Function ExceptionToMore(ByVal objException As System.Exception) As String
        Dim sb As New System.Text.StringBuilder
        With sb
            If _blnEmailError Then
                .Append("Information about this problem was automatically mailed to ")
                .Append(AppSettings.GetString("UnhandledExceptionManager/EmailTo"))
                .Append(Environment.NewLine + Environment.NewLine)
            End If
            .Append("Detailed technical information follows: " + Environment.NewLine)
            .Append("---" + Environment.NewLine)
            Dim x As String = UnhandledExceptionManager.ExceptionToString(objException)
            .Append(x)
        End With
        Return sb.ToString
    End Function

    '--
    '-- perform our string replacements for (app) and (contact), etc etc
    '-- also make sure More has default values if it is blank.
    '--       
    Private Shared Sub ProcessStrings(ByRef strWhatHappened As String, _
        ByRef strHowUserAffected As String, _
        ByRef strWhatUserCanDo As String, _
        ByRef strMoreDetails As String)

        strWhatHappened = ReplaceStringVals(strWhatHappened)
        strHowUserAffected = ReplaceStringVals(strHowUserAffected)
        strWhatUserCanDo = ReplaceStringVals(strWhatUserCanDo)
        strMoreDetails = ReplaceStringVals(GetDefaultMore(strMoreDetails))
    End Sub

    '-- 
    '-- simplest possible error dialog
    '--
    Public Overloads Shared Function ShowDialog(ByVal strWhatHappened As String, _
                                                ByVal strHowUserAffected As String, _
                                                ByVal strWhatUserCanDo As String) As DialogResult
        _blnHaveException = False
        Return ShowDialogInternal(strWhatHappened, strHowUserAffected, strWhatUserCanDo, "", MessageBoxButtons.OK, MessageBoxIcon.Warning, UserErrorDefaultButton.Default)
    End Function

    '--
    '-- advanced error dialog with Exception object
    '--
    Public Overloads Shared Function ShowDialog(ByVal strWhatHappened As String, _
        ByVal strHowUserAffected As String, _
        ByVal strWhatUserCanDo As String, _
        ByVal objException As System.Exception, _
        Optional ByVal Buttons As MessageBoxButtons = MessageBoxButtons.OK, _
        Optional ByVal Icon As MessageBoxIcon = MessageBoxIcon.Warning, _
        Optional ByVal DefaultButton As UserErrorDefaultButton = UserErrorDefaultButton.Default) As DialogResult

        _blnHaveException = True
        _strExceptionType = objException.GetType.FullName
        Return ShowDialogInternal(strWhatHappened, strHowUserAffected, strWhatUserCanDo, _
            ExceptionToMore(objException), Buttons, Icon, DefaultButton)
    End Function


    '--
    '-- advanced error dialog with More string
    '-- leave "more" string blank to get the default
    '--
    Public Overloads Shared Function ShowDialog(ByVal strWhatHappened As String, _
        ByVal strHowUserAffected As String, _
        ByVal strWhatUserCanDo As String, _
        ByVal strMoreDetails As String, _
        Optional ByVal Buttons As MessageBoxButtons = MessageBoxButtons.OK, _
        Optional ByVal Icon As MessageBoxIcon = MessageBoxIcon.Warning, _
        Optional ByVal DefaultButton As UserErrorDefaultButton = UserErrorDefaultButton.Default _
        ) As DialogResult

        _blnHaveException = False
        Return ShowDialogInternal(strWhatHappened, strHowUserAffected, strWhatUserCanDo, strMoreDetails, _
            Buttons, Icon, DefaultButton)
    End Function

    '-- 
    '-- internal method to show error dialog
    '--
    Private Shared Function ShowDialogInternal(ByVal strWhatHappened As String, _
                ByVal strHowUserAffected As String, _
                ByVal strWhatUserCanDo As String, _
                ByVal strMoreDetails As String, _
                ByVal Buttons As MessageBoxButtons, _
                ByVal Icon As MessageBoxIcon, _
                ByVal DefaultButton As UserErrorDefaultButton) As DialogResult

        '-- set default values, etc
        ProcessStrings(strWhatHappened, strHowUserAffected, strWhatUserCanDo, strMoreDetails)

        Dim objForm As New ExceptionDialog
        With objForm
            .Text = ReplaceStringVals(objForm.Text)
            .ErrorBox.Text = strWhatHappened
            .ScopeBox.Text = strHowUserAffected
            .ActionBox.Text = strWhatUserCanDo
            .txtMore.Text = strMoreDetails
        End With

        '-- determine what button text, visibility, and defaults are
        With objForm
            Select Case Buttons
                Case MessageBoxButtons.AbortRetryIgnore
                    .btn1.Text = "&Abort"
                    .btn2.Text = "&Retry"
                    .btn3.Text = "&Ignore"
                    .AcceptButton = objForm.btn2
                    .CancelButton = objForm.btn3
                Case MessageBoxButtons.OK
                    .btn3.Text = "OK"
                    .btn2.Visible = False
                    .btn1.Visible = False
                    .AcceptButton = objForm.btn3
                Case MessageBoxButtons.OKCancel
                    .btn3.Text = "Cancel"
                    .btn2.Text = "OK"
                    .btn1.Visible = False
                    .AcceptButton = objForm.btn2
                    .CancelButton = objForm.btn3
                Case MessageBoxButtons.RetryCancel
                    .btn3.Text = "Cancel"
                    .btn2.Text = "&Retry"
                    .btn1.Visible = False
                    .AcceptButton = objForm.btn2
                    .CancelButton = objForm.btn3
                Case MessageBoxButtons.YesNo
                    .btn3.Text = "&No"
                    .btn2.Text = "&Yes"
                    .btn1.Visible = False
                Case MessageBoxButtons.YesNoCancel
                    .btn3.Text = "Cancel"
                    .btn2.Text = "&No"
                    .btn1.Text = "&Yes"
                    .CancelButton = objForm.btn3
            End Select
        End With

        '-- set the proper dialog icon
        Select Case Icon
            Case MessageBoxIcon.Error
                objForm.PictureBox1.Image = System.Drawing.SystemIcons.Error.ToBitmap
            Case MessageBoxIcon.Stop
                objForm.PictureBox1.Image = System.Drawing.SystemIcons.Error.ToBitmap
            Case MessageBoxIcon.Exclamation
                objForm.PictureBox1.Image = System.Drawing.SystemIcons.Exclamation.ToBitmap
            Case MessageBoxIcon.Information
                objForm.PictureBox1.Image = System.Drawing.SystemIcons.Information.ToBitmap
            Case MessageBoxIcon.Question
                objForm.PictureBox1.Image = System.Drawing.SystemIcons.Question.ToBitmap
            Case Else
                objForm.PictureBox1.Image = System.Drawing.SystemIcons.Error.ToBitmap
        End Select

        '-- override the default button
        Select Case DefaultButton
            Case UserErrorDefaultButton.Button1
                objForm.AcceptButton = objForm.btn1
                objForm.btn1.TabIndex = 0
            Case UserErrorDefaultButton.Button2
                objForm.AcceptButton = objForm.btn2
                objForm.btn2.TabIndex = 0
            Case UserErrorDefaultButton.Button3
                objForm.AcceptButton = objForm.btn3
                objForm.btn3.TabIndex = 0
        End Select

        If _blnEmailError Then
            SendNotificationEmail(strWhatHappened, strHowUserAffected, strWhatUserCanDo, strMoreDetails)
        End If

        '-- show the user our error dialog
        Return objForm.ShowDialog()
    End Function


    '--
    '-- this is the code that executes in the spawned thread
    '--
    Private Shared Sub ThreadHandler()
        Dim smtp As New SimpleMail.SMTPClient
        Dim mail As New SimpleMail.SMTPMailMessage
        With mail
            .To = AppSettings.GetString("UnhandledExceptionManager/EmailTo")
            If _blnHaveException Then
                .Subject = "Handled Exception notification - " & _strExceptionType
            Else
                .Subject = "HandledExceptionManager notification"
            End If
            .Body = _strEmailBody
        End With
        '-- try to send email, but we don't care if it succeeds (for now)
        Try
            smtp.SendMail(mail)
        Catch e As Exception
            Debug.WriteLine("** SMTP email failed to send!")
            Debug.WriteLine("** " & e.Message)
        End Try
    End Sub

    '--
    '-- send notification about this error via e-mail
    '--
    Private Shared Sub SendNotificationEmail(ByVal strWhatHappened As String, _
        ByVal strHowUserAffected As String, _
        ByVal strWhatUserCanDo As String, _
        ByVal strMoreDetails As String)

        '-- ignore debug exceptions (eg, development testing)?
        If UnhandledExceptionManager.IgnoreDebugErrors Then
            If AppSettings.DebugMode Then Return
        End If

        Dim objStringBuilder As New System.Text.StringBuilder
        With objStringBuilder
            .Append("What happened:")
            .Append(Environment.NewLine)
            .Append(strWhatHappened)
            .Append(Environment.NewLine)
            .Append(Environment.NewLine)
            .Append("How this will affect the user:")
            .Append(Environment.NewLine)
            .Append(strHowUserAffected)
            .Append(Environment.NewLine)
            .Append(Environment.NewLine)
            .Append("What the user can do about it:")
            .Append(Environment.NewLine)
            .Append(strWhatUserCanDo)
            .Append(Environment.NewLine)
            .Append(Environment.NewLine)
            .Append("More information:")
            .Append(Environment.NewLine)
            .Append(strMoreDetails)
            .Append(Environment.NewLine)
            .Append(Environment.NewLine)
        End With
        SendEmail(objStringBuilder.ToString)
    End Sub

    Private Shared Sub SendEmail(ByVal strEmailBody As String)
        _strEmailBody = strEmailBody
        '-- spawn off the email send attempt as a thread for improved throughput
        Dim objThread As New Thread(New ThreadStart(AddressOf ThreadHandler))
        objThread.Name = "HandledExceptionEmail"
        objThread.Start()
    End Sub

End Class

File: src/main/VHD Director/HandledExceptionManager.cs
using System.Windows.Forms;
using System.Threading;
using System.Text.RegularExpressions;
using System.Configuration;
using System;
// --
// -- Generic HANDLED error handling class
// --
// -- It's like MessageBox, but specific to handled exceptions, and supports email notifications
// --
// -- Jeff Atwood
// -- http://www.codinghorror.com
// --
namespace VHD_Director
{
    public class HandledExceptionManager
    {

        private static bool _blnHaveException = false;

        private static bool _blnEmailError = true;

        private static string _strEmailBody;

        private static string _strExceptionType;

        private const string _strDefaultMore = "No further information is available. If the problem persists, contact (contact).";

        public static bool EmailError
        {
            get
            {
                return _blnEmailError;
            }
            set
            {
                _blnEmailError = value;
            }
        }

        public enum UserErrorDefaultButton
        {

            Default = 0,

            Button1 = 1,

            Button2 = 2,

            Button3 = 3,
        }

        // -- 
        // -- replace generic constants in strings with specific values
        // --
        private static string ReplaceStringVals(string strOutput)
        {
            string strTemp;
            if ((strOutput == null))
            {
                strTemp = "";
            }
            else
            {
                strTemp = strOutput;
            }
            // strTemp = strTemp.Replace("(app)", AppSettings.AppProduct);
            // strTemp = strTemp.Replace("(contact)", AppSettings.GetString("UnhandledExceptionManager/ContactInfo"));
            return strTemp;
        }

        // --
        // -- make sure "More" text is populated with something useful
        // --
        private static string GetDefaultMore(string strMoreDetails)
        {
            if ((strMoreDetails == ""))
            {
                System.Text.StringBuilder objStringBuilder = new System.Text.StringBuilder();
                // With...
                return objStringBuilder.ToString();
            }
            else
            {
                return strMoreDetails;
            }
        }

        // --
        // -- converts exception to a formatted "more" string
        // --
        private static string ExceptionToMore(System.Exception objException)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            // With...
            //if (_blnEmailError) {
            //    sb.Append(
            //    "Information about this problem was automatically mailed to ")
            //    .Append(AppSettings().GetString("UnhandledExceptionManager/EmailTo"))
            //    .Append((Environment.NewLine + Environment.NewLine));
            //}
            sb.Append(("Detailed technical information follows: " + Environment.NewLine)).Append(("---" + Environment.NewLine));
            string x = UnhandledExceptionManager.ExceptionToString(objException);
            sb.Append(x);
            return sb.ToString();
        }

        // --
        // -- perform our string replacements for (app) and (contact), etc etc
        // -- also make sure More has default values if it is blank.
        // --       
        private static void ProcessStrings(ref string strWhatHappened, ref string strHowUserAffected, ref string strWhatUserCanDo, ref string strMoreDetails)
        {
            strWhatHappened = ReplaceStringVals(strWhatHappened);
            strHowUserAffected = ReplaceStringVals(strHowUserAffected);
            strWhatUserCanDo = ReplaceStringVals(strWhatUserCanDo);
            strMoreDetails = ReplaceStringVals(GetDefaultMore(strMoreDetails));
        }

        // -- 
        // -- simplest possible error dialog
        // --
        public static DialogResult ShowDialog(string strWhatHappened, string strHowUserAffected, string strWhatUserCanDo)
        {
            _blnHaveException = false;


            return ShowDialogInternal(strWhatHappened,
                strHowUserAffected,
                strWhatUserCanDo,
                "",
                System.Windows.Forms.MessageBoxButtons.OK,
                System.Windows.Forms.MessageBoxIcon.Warning,
                HandledExceptionManager.UserErrorDefaultButton.Default);
        }

        // --
        // -- advanced error dialog with Exception object
        // --
        public static DialogResult ShowDialog(string strWhatHappened,
            string strHowUserAffected,
            string strWhatUserCanDo,
            System.Exception objException,
            MessageBoxButtons Buttons = System.Windows.Forms.MessageBoxButtons.OK,
            MessageBoxIcon Icon = System.Windows.Forms.MessageBoxIcon.Warning,
            UserErrorDefaultButton DefaultButton = UserErrorDefaultButton.Default)
        {
            _blnHaveException = true;
            // Warning!!! Optional parameters not supported
            // Warning!!! Optional parameters not supported
            // Warning!!! Optional parameters not supported
            _strExceptionType = objException.GetType().FullName;
            return ShowDialogInternal(strWhatHappened, strHowUserAffected, strWhatUserCanDo, ExceptionToMore(objException), Buttons, Icon, DefaultButton);
        }

        // --
        // -- advanced error dialog with More string
        // -- leave "more" string blank to get the default
        // --
        public static DialogResult ShowDialog(string strWhatHappened, string strHowUserAffected, string strWhatUserCanDo, string strMoreDetails, MessageBoxButtons Buttons = MessageBoxButtons.OK, MessageBoxIcon Icon = MessageBoxIcon.Warning, UserErrorDefaultButton DefaultButton = UserErrorDefaultButton.Default)
        {
            _blnHaveException = false;
            // Warning!!! Optional parameters not supported
            // Warning!!! Optional parameters not supported
            // Warning!!! Optional parameters not supported
            return ShowDialogInternal(strWhatHappened, strHowUserAffected, strWhatUserCanDo, strMoreDetails, Buttons, Icon, DefaultButton);
        }

        // -- 
        // -- internal method to show error dialog
        // --
        public static DialogResult ShowDialogInternal(string strWhatHappened, string strHowUserAffected, string strWhatUserCanDo, string strMoreDetails, MessageBoxButtons Buttons, MessageBoxIcon Icon, UserErrorDefaultButton DefaultButton)
        {
            // -- set default values, etc
            // ProcessStrings(ref strWhatHappened, ref strHowUserAffected, ref strWhatUserCanDo, ref strMoreDetails);
            return DialogResult.OK;
            // Error objForm = new Error();



            /*
           ExceptionDialog objForm = new ExceptionDialog();
           // With...
           strHowUserAffected.ActionBox.Text = strMoreDetails;
           strWhatHappened.ScopeBox.Text = strMoreDetails;
           ReplaceStringVals(objForm.Text).ErrorBox.Text = strMoreDetails;
           objForm.Text = strMoreDetails;
           // -- determine what button text, visibility, and defaults are
           // With...
           switch (Buttons) {
           }
           "&Ignore".AcceptButton = objForm.btn3;
           "&Retry".btn3.Text = objForm.btn3;
           "&Abort".btn2.Text = objForm.btn3;
           MessageBoxButtons.AbortRetryIgnore.btn1.Text = objForm.btn3;
           false.AcceptButton = objForm.btn3;
           false.btn1.Visible = objForm.btn3;
           "OK".btn2.Visible = objForm.btn3;
           MessageBoxButtons.OK.btn3.Text = objForm.btn3;
           objForm.btn2.CancelButton = objForm.btn3;
           false.AcceptButton = objForm.btn3;
           "OK".btn1.Visible = objForm.btn3;
           "Cancel".btn2.Text = objForm.btn3;
           MessageBoxButtons.OKCancel.btn3.Text = objForm.btn3;
           objForm.btn2.CancelButton = objForm.btn3;
           false.AcceptButton = objForm.btn3;
           "&Retry".btn1.Visible = objForm.btn3;
           "Cancel".btn2.Text = objForm.btn3;
           MessageBoxButtons.RetryCancel.btn3.Text = objForm.btn3;
           "&Yes".btn1.Visible = false;
           "&No".btn2.Text = false;
           MessageBoxButtons.YesNo.btn3.Text = false;
           "&Yes".CancelButton = objForm.btn3;
           "&No".btn1.Text = objForm.btn3;
           "Cancel".btn2.Text = objForm.btn3;
           MessageBoxButtons.YesNoCancel.btn3.Text = objForm.btn3;
             */
            //    }
            //}
            //// -- override the default button
            //switch (DefaultButton) {
            //    case UserErrorDefaultButton.Button1:
            //        objForm.AcceptButton = objForm.btn1;
            //        objForm.btn1.TabIndex = 0;
            //        break;
            //    case UserErrorDefaultButton.Button2:
            //        objForm.AcceptButton = objForm.btn2;
            //        objForm.btn2.TabIndex = 0;
            //        break;
            //    case UserErrorDefaultButton.Button3:
            //        objForm.AcceptButton = objForm.btn3;
            //        objForm.btn3.TabIndex = 0;
            //        break;
            //}
            //if (_blnEmailError) {
            //    SendNotificationEmail(strWhatHappened, strHowUserAffected, strWhatUserCanDo, strMoreDetails);
            //}
            //// -- show the user our error dialog
            //return objForm.ShowDialog();
            //EndFunctionEndclass Unknown {
            //}
        }

        // --
        // -- this is the code that executes in the spawned thread
        // --
        private static void ThreadHandler()
        {
            //SimpleMail.SMTPClient smtp = new SimpleMail.SMTPClient();
            //SimpleMail.SMTPMailMessage mail = new SimpleMail.SMTPMailMessage();
            //// With...
            //if (_blnHaveException) {
            //    Subject = ("Handled Exception notification - " + _strExceptionType);
            //}
            //else {
            //    Subject = "HandledExceptionManager notification";
            //}
            //Body = _strEmailBody;
            //// -- try to send email, but we don't care if it succeeds (for now)
            //try {
            //    smtp.SendMail(mail);
            //}
            //catch (Exception e) {
            //    Debug.WriteLine("** SMTP email failed to send!");
            //    Debug.WriteLine(("** " + e.Message));
            //}
        }

        // --
        // -- send notification about this error via e-mail
        // --
        private static void SendNotificationEmail(string strWhatHappened, string strHowUserAffected, string strWhatUserCanDo, string strMoreDetails)
        {
            //// -- ignore debug exceptions (eg, development testing)?
            //if (UnhandledExceptionManager.IgnoreDebugErrors) {
            //    if (AppSettings.DebugMode) {
            //        return;
            //    }
            //    System.Text.StringBuilder objStringBuilder = new System.Text.StringBuilder();
            //    // With...
            //    SendEmail(objStringBuilder.ToString());
            //}
        }

        void SendEmail(string strEmailBody)
        {
            //_strEmailBody = strEmailBody;
            //// -- spawn off the email send attempt as a thread for improved throughput
            //Thread objThread = new Thread(new ThreadStart(new System.EventHandler(this.ThreadHandler)));
            //objThread.Name = "HandledExceptionEmail";
            //objThread.Start();
        }
    }
}